test/unit/math/prim/mat/meta/scalar_type_test.cpp

```diff
@@ -1,17 +1,21 @@
 #include <stan/math/prim/mat.hpp>
 #include <gtest/gtest.h>
-#include <boost/type_traits.hpp>
+#include <test/unit/util.hpp>
+#include <vector>
 
-TEST(MetaTraits, scalar_type) {
-  using boost::is_same;
+TEST(MetaTraits, ScalarTypeMat) {
   using stan::scalar_type;
-  using Eigen::Matrix;
-  using Eigen::Dynamic;
+  using Eigen::MatrixXd;
+  using Eigen::VectorXd;
+  using Eigen::RowVectorXd;
+  using test::expect_same_type;
+  using std::vector;
 
-  stan::scalar_type<Matrix<double, Dynamic, Dynamic> >::type x = 1;
-  EXPECT_FLOAT_EQ(1, x);
+  expect_same_type<double, scalar_type<MatrixXd>::type>();
+  expect_same_type<double, scalar_type<VectorXd>::type>();
+  expect_same_type<double, scalar_type<RowVectorXd>::type>();
 
-  // hack to get value of template into Google test macro 
-  bool b3 = is_same<double, scalar_type<Matrix<double, Dynamic, Dynamic> >::type>::value;
-  EXPECT_TRUE(b3);
+  // TODO(Sean): This behavior doesn't seem to make sense prima facie
+  // Likely change this so that we would expect `double` here instead.
```

Comment:
Just commenting; I think `scalar_type` just unwraps one layer. I think there's a recursive version kicking around. But either way, this is improperly named.